packages/pigeon/lib/dart_generator.dart

```diff
@@ -252,15 +275,27 @@ void _writeFlutterApi(
                 indent.writeln('// ignore message');
                 call = 'api.${func.name}()';
               } else {
-                final String argType =
-                    _addGenericTypes(func.arguments[0].type, nullTag);
                 indent.writeln(
-                  'assert(message != null, \'Argument for $channelName was null. Expected $argType.\');',
+                  'assert(message != null, \'Argument for $channelName was null.\');',
                 );
+                const String argsArray = 'args';
                 indent.writeln(
-                  'final $argType input = (message as $argType$nullTag)$unwrapOperator;',
-                );
-                call = 'api.${func.name}(input)';
+                    'final List<Object$nullTag> $argsArray = (message as List<Object$nullTag>$nullTag)$unwrapOperator;');
+                final Iterable<String> argNames = indexMap(
+                    func.arguments, (int count, NamedType arg) => 'arg$count');
+                final Iterable<String> argTypes = func.arguments
+                    .map((NamedType e) => _addGenericTypes(e.type, nullTag));
+                enumerate(zip(argTypes, argNames),
```

Comment:
Same high-level question here. Why start with func.arguments, separate it into two iterators that you have to know go together by index, then re-zip them, rather than just doing a simple `enumerate` with string conversion? Isn't this all just:

```
enumerate(func.arguments, (int count, NamedType arg) {
  final String argType = _addGenericTypes(e.type, nullTag);
  final String argName = 'arg$count';
  [writeln code here, unchanged]
}
```

Slightly less code, no `zip`, and less mental state that the reader needs to keep (the two temp iterators).

Reply:
(I expect these comments will apply to the other languages as well; whatever resolution we get to here should be applied to them as well, but I won't leave comments there for now to avoid fracturing discussion).

Reply:
Same thing, `argNames` is used it 2 separate calculations so I don't want to duplicate the logic to calculate them and instead rely on basing everything off of one calculation stored in the `argNames` variable.

Reply:
Did you want to leave this one unchanged, rather than resolving it the same way as in the other case using `_getSafeArgumentName`?

I'm okay with it if you feel strongly about it, since in this case the two usages are entirely local, but I do think even here it would be a bit more understandable with duplicate delegation to a helper instead of the split-enumerate-zip-re-enumerate flow.

Reply:
Sure, done.